### ISSUE TITLE
Angular 11 support

### ISSUE THREAD

**Issue by mbdmardy** [just-jeb/angular-builders#908]
Hi Guys,

I would like to know your plan or timeline to release the stable version for Angular 11 support? your feedback and update much appreciated. This package really cool so we're planning to utilize it good work!. 

Thank you

**Comment by just-jeb** [comment]
Hi, thanks for your interest. I assume you're asking about `@angular-builders/custom-webpack`?  
If yes, then the blocking PR (#907) has been merge a few days ago and the stable version can actually be released even today.  
However I'd like to give it a week or so and see if someone stumbles across some issues with the new [`mergeRules`](https://github.com/just-jeb/angular-builders/blob/master/MIGRATION.MD#custom-webpack-builder).   
So in short the stable version will be released by the end of this week/beginning of the next week.  
It would be really helpful if you upgraded to the latest beta (`11.0.0-beta.2`) and gave it a try though before the stable version is released.

**Comment by jeetparikh** [comment]
hey @just-jeb I upgraded to `"^11.0.0-beta.3",` version and getting this build error
```
[error] TypeError: Cannot read property 'options' of undefined
    at /Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:186:51
    at Array.forEach (<anonymous>)
    at /Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:161:28
    at Array.map (<anonymous>)
    at mergeWithRule (/Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:138:17)
    at /Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:211:30
    at Array.forEach (<anonymous>)
    at /Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:161:28
    at Array.map (<anonymous>)
    at mergeWithRule (/Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:138:17)
    at customizeArray (/Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/index.js:121:24)
    at _joinArrays (/Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/join-arrays.js:44:50)
    at /Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/merge-with.js:32:17
    at Array.forEach (<anonymous>)
    at mergeTo (/Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/merge-with.js:31:10)
    at /Users/jeetparikh/development/angular-11/node_modules/webpack-merge/dist/merge-with.js:23:15
   ```
   
   Also do I need to update custom-webpack to version 11 in order to use `angular 11` or the previous version should work out of the box? Thanks

**Comment by jeetparikh** [comment]
An update. `beta.1` works fine.. problem starts with `beta.2`

**Comment by just-jeb** [comment]
Hey @jeetparikh, 

>Also do I need to update custom-webpack to version 11 in order to use angular 11 or the previous version should work out of the box? Thanks

Yes, you need the versions to be aligned. If you use `custom-webpack@10` with Angular 11 it will cause a lot of problems.

>An update. beta.1 works fine.. problem starts with beta.2

Makes very much sense, the breaking change of `webpack-merge` has been introduced in `beta.2`.  
Could you please create a minimal reproduction repo so that I could take a look? Thanks for your cooperation!

**Comment by mbdmardy** [comment]
Yes, @just-jeb I'm referring to this package @angular-builders/custom-webpack. That sounds good then and looking forward to this stable release. Thanks for the quick response @just-jeb  and keep up the good work. !

**Comment by jeetparikh** [comment]
hey @just-jeb yep that makes sense. But the fact that I don't use any specific merge strategy or rules.. Just the default version..

**Comment by just-jeb** [comment]
@jeetparikh Well, default rules are the ones that I defined in order to make it as similar as possible to the `smart` strategy. It doesn't mean they are right. But in order to check it properly and see if the problem is with the builder or `webpack-merge` I'm gonna need a repro.

**Comment by just-jeb** [comment]
@jeetparikh The issue happens because the original config uses `use` and you use the shortcut for this `loaders`.  
If you change it to `use` then the things should get better (still can't vouch that it will work properly, but let's see).  
Here is an issue to track in `webpack-merge`: https://github.com/survivejs/webpack-merge/issues/167.  
If switching `loaders` to `use` would work for you then we'll have to include it in the migration guide I guess.

**Comment by jeetparikh** [comment]
hey @just-jeb here is the repo with minimal reproduction of the same use
https://github.com/jeetparikh/builder-demo

Tried to use `use` instead of `loaders` but still the same error. You can try pulling the latest changes. Also now the repo uses v11.. thanks

**Comment by just-jeb** [comment]
Right, that's still happening.  
You're adding a loader, instead of changing an existing one, so in your case the solution would be passing `append` rule to `use`.  
Like this:
```
"mergeRules": {
      "module": {
        "rules": {
          "test": "match",
          "use": "append"
        },
      },
    }
```

I wonder whether I should put this as a default, since this is the most common case. This would mean that only if anyone would want to change options of an existing loader they would have to pass explicit rules.

**Comment by just-jeb** [comment]
Correction, you have to use the following rules:
```json
"mergeRules": {
      "module": {
        "rules": "append"
    }
}
```

This is because you have an `exclude` property, which is different in the original loaders I believe.

**Comment by renatoaraujoc** [comment]
I'm getting this on beta.3, any ideas?

![Captura de Tela 2020-12-16 às 13 43 51](https://user-images.githubusercontent.com/358128/102378588-bcd20900-3fa4-11eb-82cf-0ae0041eb669.png)

**Comment by just-jeb** [comment]
@renatoaraujoc what are your merge rukes and what is the expected outcome? Do you want to override the post-css plugins? Append? Could you create a reproduction please?

**Comment by jeetparikh** [comment]
> This is because you have an `exclude` property, which is different in the original loaders I believe.

hey @just-jeb thanks for the inputs. Does this mean it would still merge my config for my custom rules with the rules provided by angular cli? or append would mean my rules take precedence? Also does the order of rules then matter? Thanks

**Comment by jeetparikh** [comment]
hey @just-jeb getting the same error as @renatoaraujoc with `beta.4` as well and providing no config..  but works if provide this config suggested by you 
```
"mergeRules": {
      "module": {
        "rules": "append"
    }
}
```
but am not sure of the outcome. Thanks

**Comment by renatoaraujoc** [comment]
@just-jeb I'm just using the common tailwind css custom web pack config that you can find on the internet (that's the entire file). Where do I put those mergeRules thing?
Right now I'm stuck with beta.1 or I can't work...

Renato

**Comment by axmad22** [comment]
Painfully I have noticed by experience that Every angular update and they are frequent that the build config format changes....had to update the regret...tested _11.0.0-beta.4_ got the same errors like  @renatoaraujoc , apparently i missed the memo "mergeRules" whats that a new twist...here my old _webpack.config_ and _postcss.config_ could you please show us the new way of typing in whats suppose to be a simple **Config** file that would work 

thank you 

``` js
//webpack.config 
let sassReader;
try {
  // tslint:disable-next-line:no-implicit-dependencies
  sassReader= require('node-sass');
} catch {
  sassReader= require('sass');
}

module.exports = {
  // Fix for: https://github.com/recharts/recharts/issues/1463
  node: {
    fs: 'empty'
  },
  module: {
    rules: [
      {
        test: /\.scss$/,
        use: [
          'postcss-loader',
          {
            loader: 'sass-loader',
            options: {
              implementation: sassReader,
              sourceMap: false,
              sassOptions: {
                precision: 8
              }
            }
          }
        ]
      }
    ]
  }
};
// postcss.config
module.exports = {
  parser: 'postcss-scss',
  plugins: [
    require('tailwindcss')
  ]
}
```

**Comment by just-jeb** [comment]
Guys, there is a [migration guide](https://github.com/just-jeb/angular-builders/blob/master/MIGRATION.MD#migration-from-version-10-to-version-11) I update with each major version.   
I invest my time in doing this so I'd expect from you to read it at least.  
I'm doing my best in order to help you all, but as you can imagine I also have a life and a work and it's physically impossible to debug each one of your configs. I just don't have enough time for this.  
Version 11 introduces a breaking change, since the library it uses (`webpack-merge`) has introduced a breaking change.  
I'm trying to keep it as backward compatible as possible (I opened [4 issues](https://github.com/survivejs/webpack-merge/issues?q=is%3Aissue+author%3Ajust-jeb+) there in the last 2 days) but a breaking change is still a breaking change.  
This means that you might need to update your builder config. All that's required is to read the docs and understand how it works today.  
I'd really appreciate your help, thank you.

**Comment by just-jeb** [comment]
**All:**
If you don't want to invest any time in reading and understanding how the new auto-merge works please consider using [function config](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack#custom-webpack-config-function). This way you have full control on modifying the original config.  

@jeetparikh You still merge your rules, you just define a way it's merged. 
This config:
```ts
"mergeRules": {
      "module": {
        "rules": "append"
    }
}
```
means that the rules from your config are appended to the rules from the original config.  
What does it actually mean? As the [webpack docs](https://webpack.js.org/concepts/loaders/#loader-features) mention:
>Loaders can be chained. Each loader in the chain applies transformations to the processed resource. A chain is executed in reverse order. The first loader passes its result (resource with applied transformations) to the next one, and so forth. Finally, webpack expects JavaScript to be returned by the last loader in the chain.

So if your rule is appended to the rules of the original config, then your rule will actually be executed **first** for the matching files.   
This is also the way it worked before.  
When you say you're not sure of the outcome, what do you mean? Is your build producing the expected outcome or not?

**Comment by mbdmardy** [comment]
Thanks @just-jeb, that merge rules script works for me. I'm testing the loader now if everything works fine.

**Comment by axmad22** [comment]
Thanks @just-jeb and sorry did not see the [migration guide](https://github.com/just-jeb/angular-builders/blob/master/MIGRATION.MD#migration-from-version-10-to-version-11) . We wish you only the best in life, health and in your work which as you can see we cant function without.

**Comment by just-jeb** [comment]
@axmad22 Huh, thanks for the kind words.

**Comment by just-jeb** [comment]
@renatoaraujoc There is a bug in merge, for some reason one of the loaders is screwed in the resulting config:
```
{
    "test": "/(?:\\.ngfactory\\.js|\\.ngstyle\\.js|\\.tsx?)$/",
    "use": [
      {
        "0": "/",
        "1": "U",
        "2": "s",
        "3": "e",
        "4": "r",
        "5": "s",
        "6": "/",
        ...
      }
```
I'm looking into it, I suspect it's a problem with webpack-merge.

**Comment by mbdmardy** [comment]
Hi @just-jeb , Is @renatoaraujoc  issue related to using loaders? I build the solution without error using the custom config with loaders. Unfortunately, it seems loaders is not executed because there are no changes in the HTML file with SVG when I used markup-inline-loader.
I used this loader in custom webpack
![image](https://user-images.githubusercontent.com/46436822/102495501-c5c1e980-403b-11eb-8d7f-df49afb49c82.png)

**Comment by just-jeb** [comment]
@mbdmardy Yes, it's related to loaders, but I'm still not certain about what happens there yet. If you're not getting the error he does then probably it's not the same problem.    
Could you provide a repro for your use case please?

**Comment by just-jeb** [comment]
> @renatoaraujoc There is a bug in merge, for some reason one of the loaders is screwed in the resulting config:
> 
> ```
> {
>     "test": "/(?:\\.ngfactory\\.js|\\.ngstyle\\.js|\\.tsx?)$/",
>     "use": [
>       {
>         "0": "/",
>         "1": "U",
>         "2": "s",
>         "3": "e",
>         "4": "r",
>         "5": "s",
>         "6": "/",
>         ...
>       }
> ```
> 
> I'm looking into it, I suspect it's a problem with webpack-merge.

Indeed an issue with `webpack-merge`. [Here](There is a bug in merge, for some reason one of the loaders is screwed in the resulting config) is the one to track, hope it will be resolved soon.  

TBH I'm starting to question my decision to upgrade to the new version of `webpack-merge`.

**Comment by mbdmardy** [comment]
@just-jeb, I'm trying to transform the img HTML element with markup inline<img markup-inline> to SVG using loader. I don't issue with this config below and I already confirm that this config is merged in mergeConfigs (webpack-config-merger.js). It seems 

Config: 
 "customWebpackConfig": {
                            "path": "./webpack-custom.config.js",
                            "mergeRules": {
                                "module": {
                                    "rules": "append"
                                }
                            }
                        },

module.exports = {
    module: {
        rules: [
            {
                test: /\.html$/,
                use: [
                    {
                        loader: 'html-loader',
                        options: {
                            attrs: ['img:src'],
                            ignoreCustomFragments: [/\{\{.*?}}/],
                            minimize: false,
                            root: resolve(__dirname, 'src'),
                        },
                    },
                    {
                        loader: 'markup-inline-loader',
                        options: {
                            svgo: {
                                plugins: [
                                    {
                                        removeTitle: true,
                                    },
                                    {
                                        removeUselessStrokeAndFill: false,
                                    },
                                    {
                                        removeUnknownsAndDefaults: false,
                                    },
                                ],
                            },
                        },
                    },
                ],
            },
        ],
    },
};

Thiis the goal

**Original HTML**
<img class="icon" markup-inline height="1em" src="./_images/camera.svg" />

**Translated HTML**
<svg class="icon" markup-inline height="1em" viewBox="0 0 1024 1404.416" xmlns="http://www.w3.org/2000/svg">
  <path d="M960 440.384h-256v-128c0-35.312-28.656-64-64-64h-256c-35.344 0-64 28.688-64 64v128h-128v-64h-128v64c-35.344 0-64 28.688-64 64v704c0 35.376 28.656 64 64 64h896c35.344 0 64-28.624 64-64v-704c0-35.312-28.656-64-64-64z m-512-64h128v64h-128v-64z m448 768h-768v-576h768v576z m-384-128c106.032 0 192-85.938 192-192s-85.968-192-192-192-192 85.938-192 192 85.968 192 192 192z m0-256c35.344 0 64 28.624 64 64s-28.656 64-64 64-64-28.624-64-64 28.656-64 64-64z"/>
</svg>

**Comment by just-jeb** [comment]
@mbdmardy Try [disabling](https://github.com/just-jeb/angular-builders/issues/616#issuecomment-554356151) direct templates loading.

**Comment by mbdmardy** [comment]
Thanks @just-jeb , I'm gonna try those fix.

**Comment by axmad22** [comment]
Got to start learning webpack,but after some reading i tried some tests in the webpack.config to try to merge the rules, but it seems its not merging them here i what I tried also am not sure but the option _CustomizeRule.Merge_ in **webpack-merge** is undefined. 

``` js
const a = {
  module: {
    rules: [
      {
        test: /\.scss$/i,
        loader: "postcss-loader",
        options: {
          postcssOptions: {
            parser: 'postcss-js',
            plugins: [
              "postcss-import",
              "tailwindcss",
              "autoprefixer"
            ],
          },
        },
      },
    ],
  },
};
const b = {
  module: {
    rules: [
      {
        test: /\.scss$/i,
        use: [
          {
            loader: 'sass-loader',
            options: {
              implementation:  require('sass'),
              sourceMap: false,
              sassOptions: {
                precision: 8
              },
            },
          },
        ],
      },
    ],
  },
};

const result =  webpack_merge.mergeWithRules({
  module: {
    rules: {
      test: webpack_merge.CustomizeRule.Match,
      use: {
        loader: webpack_merge.CustomizeRule.Match,
        options: webpack_merge.CustomizeRule.Merge // its undefined 
      },
    },
  },
})(a, b);
module.exports = result; 
```

this is the value of the merge **result** i type 'merge' for options  undefined value  

``` json
{
    "module": {
        "rules": [
            {
                "test": {},
                "loader": "postcss-loader",
                "options": {
                    "postcssOptions": {
                        "parser": "postcss-js",
                        "plugins": [
                            "postcss-import",
                            "tailwindcss",
                            "autoprefixer"
                        ]
                    }
                }
            }
        ]
    }
}
```
Errors when I run script, wanted to test the append option but did not know how

``` js
An unhandled exception occurred: Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration.module.rules[12].use should be one of these:
   non-empty string | function | object { ident?, loader?, options?, query? } | function | [non-empty string | function | object { ident?, loader?, options?, query? }]
   -> Modifiers applied to the module when rule is matched
   Details:
    * configuration.module.rules[12].use[0] has an unknown property '0'. These properties are valid:
      object { ident?, loader?, options?, query? }...ect
```

**Comment by elliotleelewis** [comment]
FYI if you use TypeScript 4.1 (using the Angular 11.1 pre-releases), then you get this error message:

```
$ ng build --prod
✔ Browser application bundle generation complete.

Error: Cannot read property 'flags' of undefined


error Command failed with exit code 1.

```

Tried the same versions of everything with TS4.0 (while commenting out the lines of my app that rely on TS4.1 bugfixes) and everything compiles correctly. Also when the regular Angular builders (non custom-webpack) and TS4.1, the app builds correctly, so I'm pretty sure these builders just break when using TS4.1.

**Comment by jeetparikh** [comment]
> I'm doing my best in order to help you all, but as you can imagine I also have a life and a work and it's physically impossible to debug each one of your configs. I just don't have enough time for this.

hey @just-jeb totally understand where you are coming from. You have been totally awesome and super helpful via slack channels and resolving issues. Highly appreciate your efforts. Everyone including myself needs to be more careful and helpful to all OSS contributors as you guys do it in your free time and don't get paid for it. Thanks once again :)

> So if your rule is appended to the rules of the original config, then your rule will actually be executed **first** for the matching files.

Hey @just-jeb thank you for confirming.. with `beta4` the output is as expected and I provide this in my config. L get rid of it if you plan to default it
```
"mergeRules": {
                                "module": {
                                    "rules": "append"
                                }
                            }
```

**Comment by renatoaraujoc** [comment]
I'm just awaiting a resolution guys, I'd help fix stuff myself but I hardly know how to webpack... I'm currently using beta.1 so I can build my app without any issues.
Totally understand your concerns @just-jeb, you're being awesome dude.
My error is the same as @elliotleelewis.

**Comment by just-jeb** [comment]
@elliotleelewis thanks for bringing it up. I believe the problem is with version mismatch. See, the builder uses the `@latest` version of Angular Devkit, which means actually latest **stable**.
When you use pre-release version of Angular Devkit (`@next`) then it may collide with the version the builder uses (which still uses TS 4.0), thus giving you such an error.
If you work with `yarn` you can use `resolutions` in order to force the version of the devkit. Otherwise you should probably wait for the stable version of Angular `11.1`. 

@renatoaraujoc as I mentioned the issue is with `webpack-merge`, so I'm waiting for resolution as well. 
You're welcome to upvote the issue in `webpack-merge` or even contribute a fix - should be pretty easy.

**Comment by mbdmardy** [comment]
Hi @just-jeb it works the fix for the loaders awesome!. Everything works fine on my end for the beta 4 version.  pls, leave a comment once you release the stable version.  your awesome dude! you're the man!

**Comment by just-jeb** [comment]
@renatoaraujoc I issued a [PR](https://github.com/survivejs/webpack-merge/pull/172) that fixes the problem in `webpack-merge`. Once it's merged I'll be able to provide you with a fixed version.

**Comment by just-jeb** [comment]
@renatoaraujoc @axmad22 Please try [version `beta-5`](https://github.com/just-jeb/angular-builders/blob/master/packages/custom-webpack/CHANGELOG.md#1100-beta5-2020-12-22) and let me know if it works for you.

**Comment by just-jeb** [comment]
v11 is officially released.

**Comment by renatoaraujoc** [comment]
@just-jeb it's not working 100%!

Thank you!
